Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -158,27 +163,55 @@ Jit64::Jit64() : QuantizedMemoryRoutines(*this)
 
 Jit64::~Jit64() = default;
 
-void Jit64::AllocStack()
+void Jit64::ProtectStack()
 {
-#ifndef _WIN32
-  m_stack = static_cast<u8*>(Common::AllocateMemoryPages(STACK_SIZE));
-  Common::ReadProtectMemory(m_stack, GUARD_SIZE);
-  Common::ReadProtectMemory(m_stack + GUARD_OFFSET, GUARD_SIZE);
-#else
-  // For windows we just keep using the system stack and reserve a large amount of memory at the end
-  // of the stack.
+  if (!m_enable_blr_optimization)
+    return;
+
+#ifdef _WIN32
   ULONG reserveSize = SAFE_STACK_SIZE;
   SetThreadStackGuarantee(&reserveSize);
+#else
+  auto [stack_addr, stack_size] = Common::GetCurrentThreadStack();
+
+  const uintptr_t stack_base_addr = reinterpret_cast<uintptr_t>(stack_addr);
+  const uintptr_t stack_middle_addr = reinterpret_cast<uintptr_t>(&stack_addr);
+  if (stack_middle_addr < stack_base_addr || stack_middle_addr >= stack_base_addr + stack_size)
+  {
+    PanicAlertFmt("Failed to get correct stack base");
+    m_enable_blr_optimization = false;
+    return;
+  }
+
+  const long page_size = sysconf(_SC_PAGESIZE);
+  if (page_size < 0)
+  {
+    PanicAlertFmt("Failed to get page size");
+    m_enable_blr_optimization = false;
+    return;
+  }
+
+  const uintptr_t stack_guard_addr = Common::AlignUp(stack_base_addr + GUARD_OFFSET, page_size);
+  if (stack_guard_addr - stack_base_addr + GUARD_SIZE + MIN_UNSAFE_STACK_SIZE > stack_size)
```

Comment:
I'd check for `stack_guard_addr >= stack_middle_addr || stack_middle_addr - stack_guard_addr < GUARD_SIZE + MIN_UNSAFE_STACK_SIZE` instead.

Reply:
Oh, to take into account how full the stack already is? The stack should be very shallow at this point, but eh, I guess I can do it.